File: ios/StemDeckRemote/Playback/MiniPlayerView.swift
import SwiftUI

/// The YouTube-Music-style compact bar shown above the app's content
/// whenever a song is loaded but the full stem console has been minimized:
/// thumbnail, title, play/pause, and a thin progress line along the top
/// edge. Tapping anywhere but the play/pause button re-expands the console.
struct MiniPlayerView: View {
    let job: Job
    @ObservedObject var viewModel: PlayerViewModel
    let onExpand: () -> Void
    let onTogglePlay: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            progressLine

            HStack(spacing: 12) {
                thumbnail
                VStack(alignment: .leading, spacing: 1) {
                    Text(job.title ?? "Untitled")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(.white)
                        .lineLimit(1)
                    if let subtitle = job.subtitle {
                        Text(subtitle)
                            .font(.system(size: 11))
                            .foregroundStyle(.white.opacity(0.6))
                            .lineLimit(1)
                    }
                }
                Spacer()
                Button(action: onTogglePlay) {
                    Image(systemName: viewModel.isPlaying ? "pause.fill" : "play.fill")
                        .font(.system(size: 18, weight: .semibold))
                        .foregroundStyle(.white)
                        .frame(width: 36, height: 36)
                }
                .buttonStyle(.plain)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
        }
        .background(ConsoleTheme.surfaceContainerHigh)
        .contentShape(Rectangle())
        .onTapGesture(perform: onExpand)
    }

    private var progressLine: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Rectangle().fill(Color.white.opacity(0.15))
                Rectangle()
                    .fill(Color.white)
                    .frame(width: geo.size.width * CGFloat(viewModel.progress))
            }
        }
        .frame(height: 2)
    }

    @ViewBuilder
    private var thumbnail: some View {
        let placeholder = RoundedRectangle(cornerRadius: 6)
            .fill(ConsoleTheme.surfaceContainer)
            .overlay(Image(systemName: "music.note").foregroundStyle(.white.opacity(0.6)))

        Group {
            if let thumb = job.thumbnail, let url = URL(string: thumb) {
                AsyncImage(url: url) { phase in
                    if let image = phase.image {
                        image.resizable().aspectRatio(contentMode: .fill)
                    } else {
                        placeholder
                    }
                }
            } else {
                placeholder
            }
        }
        .frame(width: 40, height: 40)
        .clipShape(RoundedRectangle(cornerRadius: 6))
    }
}

File: android/app/src/main/kotlin/com/stemdeck/remote/playback/MiniPlayerView.kt
package com.stemdeck.remote.playback

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.MusicNote
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import coil.compose.AsyncImage
import com.stemdeck.remote.models.Job

/**
 * The YouTube-Music-style compact bar shown above the app's content whenever
 * a song is loaded but the full stem console has been minimized: thumbnail,
 * title, play/pause, and a thin progress line along the top edge. Tapping
 * anywhere but the play/pause button re-expands the console.
 */
@Composable
fun MiniPlayerView(job: Job, viewModel: PlayerViewModel, onExpand: () -> Unit, onTogglePlay: () -> Unit) {
    val isPlaying by viewModel.isPlaying.collectAsStateWithLifecycle()
    val currentTime by viewModel.currentTime.collectAsStateWithLifecycle()
    val duration by viewModel.duration.collectAsStateWithLifecycle()
    val progress = if (duration > 0) (currentTime / duration).coerceIn(0.0, 1.0) else 0.0

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(ConsoleTheme.surfaceContainerHigh)
            .clickable(onClick = onExpand),
    ) {
        Canvas(modifier = Modifier.fillMaxWidth().height(2.dp)) {
            drawRect(color = Color.White.copy(alpha = 0.15f))
            drawRect(color = Color.White, topLeft = Offset.Zero, size = Size(size.width * progress.toFloat(), size.height))
        }

        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier.size(40.dp).clip(RoundedCornerShape(6.dp)).background(ConsoleTheme.surfaceContainer),
                contentAlignment = Alignment.Center,
            ) {
                if (job.thumbnail != null) {
                    AsyncImage(model = job.thumbnail, contentDescription = null, contentScale = ContentScale.Crop, modifier = Modifier.size(40.dp).clip(RoundedCornerShape(6.dp)))
                } else {
                    Icon(Icons.Filled.MusicNote, contentDescription = null, tint = Color.White.copy(alpha = 0.6f))
                }
            }

            Column(modifier = Modifier.weight(1f).padding(start = 12.dp)) {
                Text(text = job.title ?: "Untitled", color = Color.White, fontSize = 14.sp, fontWeight = FontWeight.SemiBold, maxLines = 1)
                job.subtitle?.let {
                    Text(text = it, color = Color.White.copy(alpha = 0.6f), fontSize = 11.sp, maxLines = 1)
                }
            }

            IconButton(onClick = onTogglePlay) {
                Icon(
                    imageVector = if (isPlaying) Icons.Filled.Pause else Icons.Filled.PlayArrow,
                    contentDescription = if (isPlaying) "Pause" else "Play",
                    tint = Color.White,
                    modifier = Modifier.size(24.dp),
                )
            }
        }
    }
}
